File: AIペアプログラミング実践編_ソースコード/practical-java/session-5/Ans/src/main/java/com/example/inventory/service/InventoryService.java
package com.example.inventory.service;

import com.example.inventory.entity.Inventory;
import com.example.inventory.entity.TransactionType;
import com.example.inventory.repository.InventoryRepository;
import org.springframework.stereotype.Service;

import java.util.List;

/**
 * 在庫サービス
 */
@Service
public class InventoryService {
    private final InventoryRepository inventoryRepository;

    public InventoryService(InventoryRepository inventoryRepository) {
        this.inventoryRepository = inventoryRepository;
    }

    public Inventory findByProductId(Long productId) {
        return inventoryRepository.findById(productId)
                .orElseThrow(() -> new IllegalArgumentException("在庫情報が見つかりません: " + productId));
    }

    public Inventory updateQuantity(Long productId, TransactionType type, int quantity) {
        // 在庫を取得
        Inventory inventory = inventoryRepository.findById(productId)
                .orElseThrow(() -> new IllegalArgumentException("在庫情報が見つかりません: " + productId));

        // 数量の検証
        if (quantity <= 0) {
            throw new IllegalArgumentException("数量は1以上で入力してください");
        }

        // 在庫の更新
        int newQuantity;
        if (type == TransactionType.IN) {
            newQuantity = inventory.getQuantity() + quantity;
        } else {
            newQuantity = inventory.getQuantity() - quantity;
            if (newQuantity < 0) {
                throw new IllegalStateException("在庫が不足しています");
            }
        }

        inventory.setQuantity(newQuantity);
        return inventoryRepository.save(inventory);
    }

    public List<Inventory> findAlertItems() {
        return inventoryRepository.findAlertItems();
    }

    /**
     * アラート対象の在庫を取得（閾値以下の在庫）
     */
    public List<Inventory> getAlertProducts() {
        return inventoryRepository.findAll().stream()
                .filter(inv -> inv.getQuantity() <= inv.getThreshold())
                .toList();
    }

    /**
     * 在庫数量を更新（String型のtypeを受け取るオーバーロード）
     */
    public Inventory updateQuantity(Long productId, String type, int quantity) {
        TransactionType transactionType = "IN".equals(type) ? TransactionType.IN : TransactionType.OUT;
        return updateQuantity(productId, transactionType, quantity);
    }
}

File: AIペアプログラミング実践編_ソースコード/practical-python/session-5/exercise/inventory/services/inventory_service.py
"""在庫サービス"""
from inventory.models.inventory import Inventory
from inventory.models.transaction import TransactionType


class InventoryService:
    """在庫のビジネスロジック"""

    def __init__(self, repository=None):
        self.repository = repository

    def update_quantity(
        self,
        product_id: int,
        transaction_type: TransactionType,
        quantity: int
    ) -> Inventory:
        """在庫数量を更新"""
        # 在庫を取得
        inventory = self.repository.find_by_id(product_id)
        if inventory is None:
            raise ValueError(f"在庫情報が見つかりません: {product_id}")

        # 数量の検証
        if quantity <= 0:
            raise ValueError("数量は1以上で入力してください")

        # 在庫の更新
        if transaction_type == TransactionType.IN:
            inventory.quantity += quantity
        else:
            new_quantity = inventory.quantity - quantity
            if new_quantity < 0:
                raise ValueError("在庫が不足しています")
            inventory.quantity = new_quantity

        # 保存
        return self.repository.save(inventory)

    def get_alert_products(self):
        """アラート対象の在庫を取得"""
        # TODO: 演習 1 で実装
        raise NotImplementedError("演習 1 で実装してください")
